rfd/9999-oidc-idp-for-aws-rds-dbs-discovery.md

```diff
@@ -0,0 +1,495 @@
+  - `pairwise`: we can return a hash of the subject's identity (to hide their email, for instance)
+
+##### JSON Web Key Sets
+This endpoint returns a list of public keys (usually only one key, except for periods of key rotation).
+
+This endpoint URI must be equal to `jwks_uri` defined in the previous section.
+
+As an example, `<teleportProviderURL>/.well-known/jwks-oidc`.
+
+It should return the following:
+
+```
+200 OK
+
+{
+  "keys": [
+    {
+      "kty": "RSA",
+      "alg": "RS256",
+      "n": "<public key value>",
+      "e": "AQAB"
+    }
+  ]
+}
+```
+
+This must respect the RFC7517 (eg of a key can be found at https://www.rfc-editor.org/rfc/rfc7517#appendix-A.1).
+
+Obtaining this key should be possible using
+```golang
+GetCertAuthority(
+	ctx,
+	types.CertAuthID{
+		Type:       types.OIDCSigner, // new key type
+		DomainName: clusterName,
+	},
+	false /*loadKeys*/,
+)
+```
+
+#### AWS Role for Teleport OIDC IdP
+
+The user will create or associate an AWS Role to the Teleport OIDC IdP.
+This role must have access to list RDS Databases and have a policy that trusts Teleport as an Identity Provider.
+
+At least a policy allowing the following must be part of the role:
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": "rds:DescribeDBInstances",
+            "Resource": "*"
+        }
+    ]
+}
+```
+
+The AWS Role must trust Teleport as an IdP.
+To do so, the user must add the following Trusted relationship:
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "Federated": "arn:aws:iam::<account id>:oidc-provider/<teleportProviderURL>"
+            },
+            "Action": "sts:AssumeRoleWithWebIdentity",
+            "Condition": {
+                "StringEquals": {
+                    "<teleportProviderURL>:aud": "sts.amazonaws.com"
+                }
+            }
+        }
+    ]
+}
+```
+
+### Security
+
+#### Rotation
+JWT Signing keys can be rotated using `tctl auth rotate --type=oidc`.
+
+JWKS endpoint returns a list of public keys, so the old and the new key are provided for a seamless migration.
+
+#### MITM between AWS and Teleport
+
+The only configuration AWS has about the provider is its URL.
+If that DNS is controlled by some evil part, AWS might accept requests from that evil part.
+
+However, assuming the DNS is controlled by an evil part, then the users that are logging in to Teleport are also providing their credentials to that part.
+
+We don't think this is a scenario we should focus on this RFD.
+
+#### AWS Calls from unauthorized Teleport Users
+We'll be able to call `DescribeDBInstances` AWS endpoint, however we must ensure that this is protected by RBAC.
+
+In order to do so, we'll re-use the `db` resource with the `list` verb to allow listing RDS Databases.
```

Comment:
Picture the state where we have an Integration with AWS.
We'll have an HTTP endpoint in Proxy that calls the AWS API Describe DB Instance.
We'll have a new resource Integration, and subsequent rules (create, delete, list, update), but that only works for CRUD operations over the resource itself.

Should any user be able to list the AWS RDS DBs? Or do we want some kind of check before calling the AWS Endpoint?
I'm overloading the `db.list` rule with something else not entirely relevant.
However, I think we should add some check before calling the AWS Endpoint.

Reply:
> We'll have a new verb: `use` If the user has `integration.use` then they can issue Integration API Calls

Apologies for a naive, after-the-merge question, but what is stopping us from using one of the existing verbs like "read" for this? Does it have to do with the part where you mention rules working only for CRUD operations over the resource itself?

We have managed to avoid adding a verb like "use" for resources where one could theoretically make use of it as well ("using" a db server or a node) so I'm wondering how this situation is different.

If it's related to the `POST .../integration/:id/action/aws-oidc-list-databases` endpoint mentioned in the RFD, then we could even argue that describing access to call an integration action could be made through checking the "create" verb for a new resource called "integration action". OTOH I've seen someone mention that it's hard to implement those checks for "virtual" resources that don't actually exist on the backend.

Reply:
There's some discussion here, which provide some context
https://github.com/gravitational/teleport/pull/24256#discussion_r1163043822

> We have managed to avoid adding a verb like "use" for resources where one could theoretically make use of it as well ("using" a db server or a node) so I'm wondering how this situation is different.

They are protected by `<resource>_labels`, so we could also use that here.
We discussed that approach but ended deciding on using the `VerbUse`

As for creating a virtual resource `integration action`, I don't think it was considered.
I'm not aware of any issues of implementing resources without a backend. But I didn't try it.

I think it could be an alternative, but don't actually see a big disadvantage in using the `VerbUse`